tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/AsynchronousCSTest.java

```diff
@@ -0,0 +1,314 @@
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package org.eclipse.microprofile.fault.tolerance.tck;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+
+import javax.inject.Inject;
+
+import org.eclipse.microprofile.fault.tolerance.tck.asynchronous.AsyncClassLevelClient;
+import org.eclipse.microprofile.fault.tolerance.tck.asynchronous.AsyncClient;
+import org.eclipse.microprofile.fault.tolerance.tck.util.Connection;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.testng.Arquillian;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.testng.Assert;
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.Test;
+
+/**
+ * Verify the asynchronous invocation with COmpletionStage
+ *
+ * @author Ondro Mihalyi
+ */
+public class AsynchronousCSTest extends Arquillian {
+
+    private @Inject
+    AsyncClient client;
+
+    private @Inject
+    AsyncClassLevelClient clientClass;
+
+    private List<CompletableFuture<Void>> waitingFutures = new ArrayList<>();
+
+    @Deployment
+    public static WebArchive deploy() {
+        JavaArchive testJar = ShrinkWrap
+                .create(JavaArchive.class, "ftAsynchronous.jar")
+                .addClasses(AsyncClient.class, AsyncClassLevelClient.class, Connection.class)
+                .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml")
+                .as(JavaArchive.class);
+
+        WebArchive war = ShrinkWrap.create(WebArchive.class, "ftAsynchronous.war").addAsLibrary(testJar);
+        return war;
+    }
+
+    /**
+     * Test that the stage returned by calling an asynchronous method is not
+     * done if called right after the operation
+     */
+    @Test
+    public void testAsyncIsNotFinished() {
+        CompletableFuture<Void> waitingFuture = newWaitingFuture();
+        
+        CompletionStage<Connection> resultFuture = null;
+        resultFuture = client.serviceCS(waitingFuture);
+
+        Assert.assertFalse(resultFuture.toCompletableFuture().isDone());
```

Comment:
Note that `CompletionStage` impls need not support `toCompletableFuture()`...

Reply:
You're right, I didn't realize that. Do you have any idea how to simple find out if a CompletionStage is completed? I could create another CompletableFuture which would be completed in CompletionStage's final callback but maybe you know of a better solution?

Example of the solution I have in mind:

```
CompletableFuture resultFutureStatus = new CompletableFuture();
resultFuture.whenComplete( (v, e) -> {
 if (e != null) {
  resultFutureStatus.completeExceptionally(e);
 } else {
  resultFutureStatus.complete(null);
 }
}

Assert.assertFalse(resultFutureStatus.isDone());
```

Reply:
Well, I think you could make use of some synchronization aid such as `java.util.concurrent.CountDownLatch`, e.g. something like:
````java
AtomicBoolean completedNormally = new AtomicBoolean(false);
CountDownLatch latch = new CountDownLatch(1);
foo.someAsyncMethod().whenComplete((v, e) -> {
  latch.countDown();
  if(e == null) {
    completedNormally.set(true);
  }
latch.await(500, TimeUnit.MILLISECONDS);
Assert.assertTrue(completedNormally.get());
````
And the `AtomicBoolean` is unnecessary if you only want to test whether a stage is complete or not.

Reply:
I solved this using a CompletableFuture instead of CuntDownLatch and AtomicBoolean, I think it's simpler. Please review if it's correct.

Reply:
Looks good but I did not have enough time to go through all the tests and details...